src/nifs/protogalaxy/tests.rs

```diff
@@ -0,0 +1,134 @@
+use halo2_proofs::{
+    halo2curves::{
+        ff::{FromUniformBytes, PrimeFieldBits},
+        group::prime::PrimeCurveAffine,
+    },
+    plonk::Circuit,
+};
+
+use super::{super::tests::*, *};
+
+const T: usize = 3;
+const RATE: usize = 2;
+const R_F: usize = 4;
+const R_P: usize = 3;
+
+use crate::{
+    halo2curves::bn256::G1Affine as Affine,
+    nifs::tests::random_linear_combination_circuit::RandomLinearCombinationCircuit,
+    poseidon::{PoseidonHash, Spec},
+    table::{CircuitRunner, Witness},
+};
+
+type Scalar = <Affine as CurveAffine>::ScalarExt;
+type Base = <Affine as CurveAffine>::Base;
+
+type RO<F> = PoseidonHash<F, T, RATE>;
+type Instance<F> = Vec<F>;
+const L: usize = 2;
+
+type ProtoGalaxy = crate::nifs::protogalaxy::ProtoGalaxy<Affine, L>;
+type ProverParam = <ProtoGalaxy as FoldingScheme<Affine, L>>::ProverParam;
+type VerifierParam = <ProtoGalaxy as FoldingScheme<Affine, L>>::VerifierParam;
+
+struct Mock<CIRCUIT: Circuit<Scalar>> {
+    S: PlonkStructure<Scalar>,
+    ck: CommitmentKey<Affine>,
+
+    ro_nark_verifier: RO<Base>,
+    ro_acc_prover: RO<Base>,
+    ro_acc_verifier: RO<Scalar>,
+
+    circuit_meta: [(Witness<Scalar>, Instance<Scalar>); L],
+
+    pp: ProverParam,
+    vp: VerifierParam,
+
+    _p: PhantomData<CIRCUIT>,
+}
+
+impl<C: Circuit<Scalar>> Mock<C> {
+    pub fn new(k_table_size: u32, circuits: [(C, Vec<Scalar>); L]) -> Self {
+        let circuits_runners =
+            circuits.map(|(circuit, instance)| CircuitRunner::new(k_table_size, circuit, instance));
+
+        let ck = setup_smallest_commitment_key(k_table_size, &circuits_runners[0].cs, b"");
+        let S = circuits_runners[0]
+            .try_collect_plonk_structure()
+            .expect("failed to collect plonk structure");
+
+        let circuit_meta = circuits_runners.map(|cr| {
```

Comment:
circuit_witness is better, meta might related to circuit meta info which is confused with circuit structure

Reply:
Will add `CircuitCtx` struct